### DISCUSSION TITLE
How does window size downscale drawings?

### DISCUSSION THREAD

**Discussion by drum445** [hajimehoshi/ebiten#2671] [Q&A]
Does drawing sprites to the screen at 2x the desired size then calling `SetWindowSize` with windowSize/2 produce the exact same image as simply drawing to the screen at the desired tile size?

I have used this approach to prevent blurry text when maximising the window, so I draw everything to my screen at 2x the size, then half the window width and height in SetWindowSize. So if the user does maximise the window, nothing will actually be stretched

Is that the recommended approach for ebiten to avoid blurry images/text when maximising the window or when running on high-dpi screens?

**Comment by hajimehoshi** [comment]
Have you looked at `examples/highdpi`? Probably this is what you want to know. This example implements `Layout` for hi-DPI environments not to blur renderings.

**Reply by drum445** [reply]
I did thanks, but that seems like it would require too many code change to prevent blurry text. I will take another loo, but do you think my approach seems reasonable? It seems to have done the trick, but want to make sure I'm not overlooking something

On that highdpi example, I couldn't see it being applied to drawing text?

**Reply by hajimehoshi** [reply]
I am not sure your approach. What is the 'screen' you meant? Is this a big offscreen?

> On that highdpi example, I couldn't see it being applied to drawing text?

Right, but the approach should be the same.

**Reply by drum445** [reply]
Sorrry, I probably wasn't very clear.
For example, previously if I wanted to draw some text that is 20px in size, I would just create a font object like so:
```
opentype.NewFace(tt, &opentype.FaceOptions{
		Size:    20,
		DPI:     72,
		Hinting: font.HintingVertical,
	})
```
and pass that into text.Draw(...)

This worked fine when the window was at the size I intend, let's say 480p. However, I want to allow the user to maximise the window, but when this happened, the text would stretch and become blurry as it was still only being drawn at 20px then stretched to fill the maximised window.

My solution to this was to instead double my windowWidth/Height, e.g. 480*2, draw the text at 40px instead of 20px, then at my program start I would simply do `ebiten.SetWindowSize(windowWidth/2, windowHeight/2)`, which would mean the window would still correctly render at 480p on load, but now when the user maximises the window, nothing is really stretched, instead the text is now just drawn at its original 40px again.

In short, I've switched it from initially rendering at full size, then stretching during maximising, to initially rendering at half size, then returning to original size when maximising

Does that make sense?

**Reply by hajimehoshi** [reply]
I'm afraid I don't understand your idea yet. Could you give an entire code?

**Reply by hajimehoshi** [reply]
I feel like changing the window size for a finer rendering would not make sense. This is Layout's responsibility actually.

**Reply by drum445** [reply]
I've put this example together

```
package main

import (
	"fmt"
	"image/color"
	"log"

	"github.com/hajimehoshi/ebiten/examples/resources/fonts"
	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/text"
	"golang.org/x/image/font"
	"golang.org/x/image/font/opentype"
)

const (
	windowWidth  = 1920
	windowHeight = 1080
)

var (
	mplusSmallerFont font.Face
)

type Game struct{}

func (g *Game) Update() error {
	return nil
}

func (g *Game) Draw(screen *ebiten.Image) {
	screen.Fill(color.White)
	text.Draw(screen, "SOME EXAMPLE TEXT", mplusSmallerFont, 100, 100, color.Black)
}

func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
	return windowWidth, windowHeight
}

func main() {

	tt, _ := opentype.Parse(fonts.MPlus1pRegular_ttf)
	mplusSmallerFont, _ = opentype.NewFace(tt, &opentype.FaceOptions{
		Size:    80,
		DPI:     72,
		Hinting: font.HintingVertical,
	})

	ebiten.SetWindowSize(windowWidth/2, windowHeight/2)
	ebiten.SetWindowTitle("Example")
	ebiten.SetWindowResizingMode(ebiten.WindowResizingModeEnabled)
	ebiten.SetTPS(60)

	fmt.Println(windowWidth/2, windowHeight/2)
	if err := ebiten.RunGame(&Game{}); err != nil {
		log.Fatal(err)
	}
}
```

If you run that you'll see that the font is rendered nicely with no blurring even when increasing the resolution
This is because the font is actually being drawn at 80px, but as I half the window size it is scaled down to 40px which means it can go above 40 without stretching occurring. To see this more clearly change the following lines and you'll see what I mean:

`windowWidth  = 960`
`windowHeight = 540`
`Size:    40,` 
`ebiten.SetWindowSize(windowWidth, windowHeight)`

**Reply by hajimehoshi** [reply]
Isn't this the same as

```go
const (
	windowWidth  = 960
	windowHeight = 540
)

func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
	return windowWidth*2, windowHeight*2
}

func main() {
	// ...
	ebiten.SetWindowSize(windowWidth, windowHeight)
	// ...
}
```

So this is exactly the same as what the `highdpi` example does, but your idea always uses x2 scaling. This might be overkill when a display is not hi-dpi (where `ebiten.DeviceScaleFactor()` < 2).

**Reply by drum445** [reply]
In theory yes, but I'd still have to draw the font twice as big right? Or more accurately `*ebiten.DeviceScaleFactor()`

**Reply by hajimehoshi** [reply]
> but I'd still have to draw the font twice as big right? Or more accurately *ebiten.DeviceScaleFactor()

Yes.

**Reply by drum445** [reply]
Thanks for your replies, I'll give that a go and compare the differences between the two approaches.
It' a little more complex than my example, as I have many different entities drawing to the screen, so just trying to find the simplest option
